cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java

```diff
@@ -0,0 +1,110 @@
+import io.cdap.cdap.app.runtime.ProgramController;
+import io.cdap.cdap.app.runtime.ProgramRuntimeService;
+import io.cdap.cdap.app.runtime.ProgramRuntimeService.RuntimeInfo;
+import io.cdap.cdap.app.store.Store;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.AbstractRetryableScheduledService;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.app.store.RunRecordDetail;
+import io.cdap.cdap.proto.ProgramRunStatus;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+
+/**
+ * A service that periodically scans for program runs that are in Stopping state and
+ * force terminates them if they are running beyond the expected terminateTime.
+ */
+public class ProgramRunStatusMonitorService extends AbstractRetryableScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunStatusMonitorService.class);
+
+  private final Store store;
+  private final ProgramRuntimeService runtimeService;
+  private final int txBatchSize;
+  private final long intervalMillis;
+
+  @Inject
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService) {
+    this(cConf, store, runtimeService, cConf.getInt(Constants.AppFabric.PROGRAM_TERMINATOR_TX_BATCH_SIZE),
+         cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_SECS));
+  }
+
+  @VisibleForTesting
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize, long specifiedIntervalSecs) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    if (specifiedIntervalSecs <= 0) {
+      LOG.warn("Invalid interval {} specified for the program terminator. Setting it to 5 minutes.",
+               specifiedIntervalSecs);
+      specifiedIntervalSecs = TimeUnit.MINUTES.toMillis(5);
+    }
+    this.intervalMillis = TimeUnit.SECONDS.toMillis(specifiedIntervalSecs);
+  }
+
+  @Override
+  protected long runTask() {
+    terminateStoppingProgramsActiveBeyondTerminateTime();
+    return this.intervalMillis;
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
+    // fetch all runs that are in stopping state that started at most a minute before current time.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    long currentTimeInMillis = System.currentTimeMillis();
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(
```

Comment:
This has a risk of consuming too much memory. Better fetch in batches

Reply:
Was already fetching in batches. Using a filter now to filter more runs based on terminateTime